src/io/dataset.cpp

```diff
@@ -690,6 +707,51 @@ void Dataset::SaveBinaryFile(const char* bin_filename) {
   }
 }
 
+void Dataset::DumpTextFile(const char* text_filename){
+  auto file = fopen(text_filename, "wt");
+  fprintf(file, "num_features: %d\n", num_features_);
+  fprintf(file, "num_total_features: %d\n", num_total_features_);
+  fprintf(file, "num_groups: %d\n", num_groups_);
+  fprintf(file, "num_data: %d\n", num_data_);
+  fprintf(file, "feature_names: ");
+  for(auto n : feature_names_){
+    fprintf(file, "%s, ", n.c_str());
+  }
+  fprintf(file, "\nmonotone_types: ");
+  for(auto i : monotone_types_){
+    fprintf(file, "%d, ", i);
+  }
+  fprintf(file, "\nfeature_penalty: ");
+  for(auto i : feature_penalty_){
+    fprintf(file, "%lf, ", i);
+  }
+  fprintf(file, "\n");
+  for(auto n : feature_names_){
+    fprintf(file, "%s, ", n.c_str());
+  }
+  std::vector<std::unique_ptr<BinIterator>> iterators(num_total_features_);
+  for(int j = 0; j < num_total_features_; j++){
+    if(used_feature_map_[j] >= 0){
+	auto real_feature_idx = used_feature_map_[j];
+	auto group_idx = feature2group_[real_feature_idx];
+	auto sub_idx = feature2subfeature_[real_feature_idx];
+	std::unique_ptr<BinIterator> sub_iter(feature_groups_[group_idx]->SubFeatureIterator(sub_idx));
+	iterators[j] = std::move(sub_iter);
```

Comment:
move or copy?

Reply:
Move seems OK here, the iterator is allocated on line 738, we want to move ownership of it to the iterators vector.
@guolinke, do you agree?